### PR TITLE
DDF-2287: Added integration tests for download interrupt scenarios.

### PR DESCRIPTION
**Pull request by Lambeaux**
#### What does this PR do?
Adds a set of integration tests to the `TestFederation` class for ensuring interruptions to parallel downloads do not cause corruption of product retrieval. 

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?

@garrettfreibott 
@oconnormi 
@cmthom10 
@pklinef 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).

@figliold 
@shaundmorris

#### How should this be tested?
Validate a full build and that `TestFederation` completely passes. 

#### Any background context you want to provide?
These are tests to provide a reliable harness for tweaking resource download code in the standard framework. 

#### What are the relevant tickets?
DDF-2287

##### Change set

* Created the `ClientDownloader` classes for parallel integration testing
* Updated async http libraries in test
* Fix pom dependency to OSGI version

##### Details

Within `TestFederation`, the actual test method is `runParallelInterruptDownloadTest(...)` which can be configured with the caller's choice of objects implementing `ClientDownloader`, which is the base used to abstract the details of how a particular client can be cancelled, interrupted, or abruptly closed. These classes can be found in `ddf.test.itests.common.downloads` and future implementations should live there as well. 

The actual test methods call variations of `runParallelInterruptDownloadTest(...)` and the current state of the test matrix can be found in the next section. Note that some methods are included but are labeled as `@Ignore`, which are necessary test cases but cannot currently be implemented. These cases are also in the matrix labeled as such. 

The `org.apache.http` async client libraries were necessary so that downloads could actually be started and continue to progress in parallel. Blocking calls would render the tests worthless. Optionally, a custom stream consumer could be injected into the client to provide custom callback logic for handling characters as they are received. This strategy was attempted with the creation of `InterruptableAsyncCharConsumer` but failed when _the underlying `IOControl` instance was redundant to just closing the original client object_ and catching the exception. Furthermore, once an event was fired and the consumer started working, it would continue to the end of the stream and always yield a valid cancelled product. 

As it stands, _this custom consumer is pending removal from this PR_ without replacement, but that means the tests will never know the degree of progress made on the partial product before interruption because _abruptly closing the socket means forfeiting the resultant data_. 

These circumstances might be acceptable for this set of tests and is pending feedback. Source has been pre-commented for key responses that are needed. 

##### Parallel Test Matrix

Case | Description | Client 1 | Client 2 | Flip Start Order
---- | ----- | ----- | ----- | -----
1 | Two synchronous clients and one shuts down | Synchronous | Synchronous | N/A (Negligible)
2 | Async starts followed by sync and sync shuts down | Synchronous | Asynchronous | false
3 | Sync starts followed by async and sync shuts down | Synchronous | Asynchronous | true
4 | Async starts followed by sync and async cancels (TODO) | Asynchronous | Synchronous | true
5 | Sync starts followed by async and async cancels (TODO) | Asynchronous | Synchronous | false
6 | Two asynchronous clients start and one cancels (TODO) | Asynchronous | Asynchronous | N/A (Negligible)

##### Pending

* Implementation of asynchronous cancellation (another ticket TBD)

### CODE REVIEW

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContent.java [hunk 1]
```diff
@@ -130,6 +131,12 @@ public Response apply(Response response) {
         private Response respond(Response response) {
             for (char c : responseMessage.toCharArray()) {
                 try {
+                    String log = String.format("[2287] Sending Character: [%s] at %s%n%n",
+                            String.valueOf(c),
+                            DateTime.now()
+                                    .toString());
+                    System.out.println(log);
```
Comment (shaundmorris) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContent.java]: Why do you have a System.out.println here? Should the logger be used?

Comment (Lambeaux) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContent.java]: For some reason the logger statements were not showing up. Do tests currently default to debug or do I need to pass a parameter?

Comment (tbatie) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContent.java]: Logging is turned off by default to save time during tests. You can see them if you use the flag -DitestLogLevel=DEBUG

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ParallelConsumer.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * Useful for processing Restito responses safely in parallel, with additional effects if desired.
+ */
+public class ParallelConsumer {
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ParallelConsumer.java]: This class will be getting deleted. Forgot I made it. This was back when I was trying to work with blocking calls. You may disregard.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 3]
```diff
@@ -1181,7 +1182,8 @@ public void testRetrievalReliablility() throws Exception {
                         Condition.parameter("request", "GetRecordById"),
                         Condition.parameter("id", metacardId));
 
-        // Add CometD notification and activity assertions when DDF-2272 ihas been addressed.
+        // TODO: Add CometD notification and activity assertions when DDF-2272 ihas been addressed.
+        //      Does this need to be addressed?
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: What's the status on this?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 4]
```diff
@@ -1683,13 +1718,97 @@ private CometDClient setupCometDClient(List<String> channels) throws Exception {
...
+        logParallelInterruptTestState("[2287]", resourceData, result1, result2);
+        assertThat("Valid result was not complete", result2, is(resourceData));
+    }
+
+    private void logParallelInterruptTestState(String prefix, String... args) {
+        Iterator<String> arg = Arrays.asList(args)
+                .iterator();
+        System.out.println(String.format("%s Resource Data: %s", prefix, arg.next()));
+        System.out.println(String.format("%s Result 1: %s", prefix, arg.next()));
+        System.out.println(String.format("%s Result 2: %s", prefix, arg.next()));
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: I can fix all of this later. I was just not seeing any output from `LOGGER.info(...)` so I defaulted to this so I could make progress.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 5]
```diff
@@ -1683,13 +1718,97 @@ private CometDClient setupCometDClient(List<String> channels) throws Exception {
...
+    }
+
     @Override
     protected Option[] configureCustom() {
-        return options(mavenBundle("ddf.test.thirdparty", "restito").versionAsInProject());
+        return combineOptions(options(mavenBundle("ddf.test.thirdparty",
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Figuring this statement out took most of this ticket's time. The new dependencies I introduced had _other_ dependencies that weren't being found by the container. All of these are needed.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/AsynchronousClientDownloader.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                        null);
+
+        try {
+            HttpResponse response = (BasicHttpResponse) future.get();
+            return IOUtils.toString(response.getEntity()
+                    .getContent(), "UTF-8");
+        } catch (InterruptedException e) {
+            Thread.currentThread()
+                    .interrupt();
+            return processResults();
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/AsynchronousClientDownloader.java]: This is probably horrible code because it, theoretically, could result in an infinite loop if my thread keeps getting interrupted. The odds are astronomical, and given this is test scope, I was OK with it. But I would really like @pklinef's feedback on the matter. It was the most elegant way to address the fact that _"regardless of interrupt, I want you to keep doing what I want you to do"_. Thoughts?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    public abstract void startDownload(String metacardId);
+
+    public abstract void stopDownload();
+
+    public CloseableHttpAsyncClient getClient() {
+        return client;
+    }
+
+    public boolean hasException() {
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java]: Do we want to add an assertion on this method in the test code since we always expect a `ConnectionClosedException` wrapped in an `ExecutionException` for the cancelled client?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/InterruptableAsyncCharConsumer.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.http.HttpResponse;
+import org.apache.http.nio.IOControl;
+import org.apache.http.nio.client.methods.AsyncCharConsumer;
+import org.apache.http.protocol.HttpContext;
+
+/**
+ * Socket ingest strategy for testing the download of products.
+ */
+public class InterruptableAsyncCharConsumer extends AsyncCharConsumer<String> {
+    private StringBuilder builder = new StringBuilder();
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/InterruptableAsyncCharConsumer.java]: If we don't want to cache the results from the closed client using some form of consumer in `org.apache.http` then this can be removed.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/SynchronousClientDownloader.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.apache.http.nio.client.methods.HttpAsyncMethods;
+
+/**
+ * Provides an asynchronous handle to the synchronous retrieval and download endpoint of a product.
+ */
+class SynchronousClientDownloader extends ClientDownloader {
+    private final InterruptableAsyncCharConsumer consumer = new InterruptableAsyncCharConsumer();
+
+    private final StringBuilder resultsBuilder = new StringBuilder();
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/SynchronousClientDownloader.java]: If we don't want to cache the results from the closed client using some form of consumer in `org.apache.http` then this can be removed.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/SynchronousClientDownloader.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    protected String processResults() throws ExecutionException {
+        if (future == null) {
+            throw new IllegalStateException("Download was never started, so no future exists");
+        }
+        try {
+            return (String) future.get();
+        } catch (InterruptedException e) {
+            Thread.currentThread()
+                    .interrupt();
+            return processResults();
```
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/SynchronousClientDownloader.java]: Once again, I think this is OK for test scope, but I want to be sure @pklinef (same question as above).

File: distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContent.java (hunk 1 above)
Comment (garrettfreibott) [inline distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/restito/ChunkedContent.java]: Probably best to just revert the changes to this file. The PR I have up right now adds some major cleanup to this class, and one of them is this exact character log as a `LOGGER.debug` statement.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 11]
```diff
@@ -1683,13 +1718,97 @@ private CometDClient setupCometDClient(List<String> channels) throws Exception {
...
+                        contentType("text/xml"),
+                        bytesContent(getCswQueryResponse(metacardId).getBytes()));
+
+        cswServer.whenHttp()
+                .match(Condition.get("/services/csw"),
+                        Condition.parameter("request", "GetRecordById"),
+                        Condition.parameter("id", metacardId))
+                .then(getCswRetrievalHeaders(filename),
+                        chunkedContentWithHeaders(resourceData, Duration.ofMillis(50), 0));
+    }
```
Comment (cmthom10) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Lines 1727-1740 could potentially be replaced by a DownloadHandle object of the class I added to TestFederation in my PR: https://github.com/codice/ddf/pull/1009

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/AsynchronousClientDownloader.java (hunk 6 above)
Comment (cmthom10) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/AsynchronousClientDownloader.java]: Not sure how possible this is since I don’t know this code and its uses very well, but have you considered using a “number of interrupts” variable in processResults? That variable could have 1 added to it each time processResults is called because of an interruption (and it would start at 0). When it reaches a large number that probably indicates it will interrupt indefinitely, you can tell it to stop calling itself.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 12]
```diff
@@ -1672,7 +1707,7 @@ private Action getCswRetrievalHeaders(String filename) {
     }
 
     private String getResourceData(String metacardId) {
-        return String.format("Data for metacard ID %s", metacardId);
+        return String.format("Hello there. This is data for metacard ID %s", metacardId);
```
Comment (cmthom10) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Is the "Hello there" to make it easier to search for this string?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 11 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Your `DownloadHandle` is for starting downloads with Restito logic. I can't use Restito for my use case. All I need is to program the CSW server.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 12 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: To make the data longer. I needed more characters.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/AsynchronousClientDownloader.java (hunk 6 above)
Comment (pklinef) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/AsynchronousClientDownloader.java]: You should stop processing and throw an ExecutionException.  Most likely if you are being interrupted, the system is trying to shutdown.  We do not want to hang the system shutdown on a large download.

http://www.ibm.com/developerworks/library/j-jtp05236/index.html

**Comment by Lambeaux** [comment]
@shaundmorris and @figliold ... along with @garrettfreibott, @cmthom10 and @oconnormi can I get some thumbs up on this PR? Updates are in!

**Comment by bantillo** [comment]
👍

**Comment by cmthom10** [comment]
:+1: Looks good to me. I still think that your setupParallelInterruptDownloadTest could be replaced by creating a DownloadHandle when all of our tests get merged together (all that stuff is in the constructor of DownloadHandle). Or the other way around, I suppose, if it ends up making more sense to have your setup called by the DownloadHandle constructor for my code to use.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 13]
```diff
@@ -1746,9 +1771,8 @@ private Action getCswRetrievalHeaders(String filename) {
                 header("Content-Disposition", "filename=" + filename));
     }
 
-
     private String getResourceData(String metacardId) {
-        return String.format("Data for metacard ID %s", metacardId);
+        return String.format("Hello there. This is data for metacard ID %s", metacardId);
```
Comment (shaundmorris) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Why did you add `hello there`?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            Throwable causeOfException = e.getCause();
+            assertThat("Execution failed for a reason other than the connection being closed",
+                    causeOfException instanceof ConnectionClosedException,
+                    Matchers.is(Boolean.TRUE));
+            return causeOfException.toString();
+        }
+    }
+
+    public static String getSyncDownloadUrl(String metacardId) {
+        return AbstractIntegrationTest.SERVICE_ROOT.getUrl() + "/catalog/sources/"
```
Comment (shaundmorris) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java]: Shouldn't you be using String.format for these methods ?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 13 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Needed more characters in that dataset.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java (hunk 14 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java]: Yes. It looks a lot better. Thanks.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            Throwable causeOfException = e.getCause();
+            assertThat("Execution failed for a reason other than the connection being closed",
+                    causeOfException instanceof ConnectionClosedException,
+                    Matchers.is(Boolean.TRUE));
+            return causeOfException.toString();
+        }
+    }
+
+    public static String getSyncDownloadUrl(String metacardId) {
+        //        return AbstractIntegrationTest.SERVICE_ROOT.getUrl() + "/catalog/sources/"
```
Comment (shaundmorris) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java]: Should we remove the commented out return statements? I don't see any value in keeping them there.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 12 above)
Comment (shaundmorris) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: I would at least put a comment in there that you are trying to get to a specific amount of characters and the reason why. Otherwise, we could come up with a better test message. For example we can include "Mocked resource" or something along those lines in there.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 16]
```diff
@@ -1966,14 +1952,105 @@ private CometDClient setupCometDClient(List<String> channels) throws Exception {
...
+            client1.startDownload(metacardId);
+            Thread.sleep(startDelay);
+            client2.startDownload(metacardId);
+        } else {
+            client2.startDownload(metacardId);
+            Thread.sleep(startDelay);
+            client1.startDownload(metacardId);
+        }
+
+        Thread.sleep(600);
```
Comment (shaundmorris) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Why do you need the sleep here? Seems like there are a few sleeps in here. Concerned about the overall time to run the tests.

Comment (brendan-hofmann) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Thread sleeps generally lead to unreliable tests in the long run as well. I would also like to see these replaced with concrete waits until an expected state is reached.

If that is not possible, I think we should at least be using a constant instead of these magic numbers.

Comment (peterhuffer) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: [WaitCondition](https://github.com/codice/ddf/blob/master/distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/WaitCondition.java) may be useful here and allows waiting a limited time until an expected condition is met.

Comment (peterhuffer) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: [WaitCondition](https://github.com/codice/ddf/blob/master/distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/WaitCondition.java) may be useful here and allows waiting a limited time until an expected condition is met.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java (hunk 15 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/ClientDownloader.java]: Good catch. I forgot those were there.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 16 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: This was discussed prior to implementation. It's not the _best_ way to do it but priority was placed on getting tests done so that standard framework code can be safely modified ASAP.

Take a look at `ChunkedContent.java` and you'll see we have facilities in place for slowing down transfer speeds so that download tests consistently meet the criteria. If we are going to make a swap to `WaitCondition` that is a major change that would have to go on a separate ticket. It would require changing the entire infrastructure of download tests.

**Comment by Lambeaux** [comment]
Everything is set. Hero passed. Tests pass locally 👍 
@shaundmorris let me know if we're holding off on this one. It's stable so I don't see why we would.

**Comment by Lambeaux** [comment]
@shaundmorris Scratch that last comment. After talking with @stustison I can see why this would be a concern. Working on a fix ASAP.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 16 above)
Comment (garrettfreibott) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: I talked with you yeserday about this possible solution, but using Apache's `HttpClient` in those client threads to manually open an input stream from the get request, have a looping read until a specific middle-of-the-data character is read, and then making it terminate the connection is a way to make it more deterministic. According to the http spec, chunked content responses indicate the end of data content by sending a null character, so a read loop along [these lines](http://www.vogella.com/tutorials/ApacheHttpClient/article.html#example_get) should work for reading in data character-by-character as it's slowly sent by the stub server.

Comment (garrettfreibott) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Looking into more actually, HttpClient might block until the inputstream is read entirely. Might want to try logging each character read and make sure that it happens right after the server logs that it sent a character. If it ends up blocking, java.net's HttpURLConnection might be the way to go. Same concept with that.

**Comment by Lambeaux** [comment]
Closing this PR and will re-open after I rewrite the solution.
